subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/PropertyAssignmentIntegrationTest.groovy

```diff
@@ -305,6 +305,54 @@ class PropertyAssignmentIntegrationTest extends AbstractIntegrationSpec {
         "FileCollection += Iterable<File>" | "+="      | "ConfigurableFileCollection" | 'listOf(file("a.txt"))'            | unsupportedWithDescription("Val cannot be reassigned")
     }
 
+    def "test Groovy lazy property assignment with NamedDomainObjectContainer"() {
+        given:
+        buildFile """
+            class PluginDeclarationPlugin implements Plugin<Project> {
+                void apply(Project project) {
+                    NamedDomainObjectContainer<PluginDeclaration> pluginDeclarations = project.container(PluginDeclaration)
+                    project.extensions.add('pluginDeclarations', pluginDeclarations)
+                    def projectPath = project.layout.projectDirectory.asFile.absolutePath
+                    project.tasks.register("printDeclarations") {
+                        pluginDeclarations.all {
+                            println("Plugin: id=\${it.id.get()}, description=\${it.description.get()}, tags=\${it.tags.get()}, files=\${it.myFiles.files.collect { it.path.replace(projectPath, '').replace('\\\\', '/') } }")
+                        }
+                    }
+                }
+            }
+
```

Comment:
💭 IIUC, you might be able to simplify this by putting everything into the build script instead of applying a plugin like this.  You could also make an interface that has a getter for a NamedDomainObjectContainer vs creating it manually.  I initially thought it was meaningful that you were creating this manually, but based on what you said elsewhere, this happens in the java-gradle-plugin too.

Instead of printing this out in a task, you could also just use `assert pluginDeclarations.myPlugin.id == "my-id"` in the buildscript after it should have been set. I find that usually a little easier to read than bouncing between where something is set, the task that prints it out and the expectations at the end of the test.

Reply:
I can simplify it.

Although I see that if I use getter with an interface, then we decorate the object and `=` works. While when we instantiate it manually, then it the same problem as with the `gradlePlugin {}`

Reply:
That may mean we're missing some decoration down in `project.container(...)`?  That sounds like a bug